js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+      return dataView.getFloat32(offset, true);
+    }
+    return dataView.getFloat64(offset, true);
+  }
+
+  function indirect(dataView, offset, width) {
+    const step = readInt(dataView, offset, width);
+    return offset - step;
+  }
+
+  function keyIndex(key, dataView, offset, parentWidth, byteWidth, length) {
+    const input = toUTF8Array(key);
+    const keysVectorOffset = indirect(dataView, offset, parentWidth) - byteWidth * 3;
+    const bitWidth = flexbuffers.BitWidth.fromByteWidth(byteWidth);
+    const indirectOffset = keysVectorOffset - readInt(dataView, keysVectorOffset, bitWidth);
+    const _byteWidth = readInt(dataView, keysVectorOffset + byteWidth, bitWidth);
+    let low = 0;
+    let high = length - 1;
+    while (low <= high) {
+      const mid = (high + low) >> 1;
+      const dif = diffKeys(input, mid, dataView, indirectOffset, _byteWidth);
+      if (dif === 0) return mid;
+      if (dif < 0) {
+        high = mid - 1;
+      } else {
+        low = mid + 1;
+      }
+    }
+    return null;
+  }
+
+  function diffKeys(input, index, dataView, offset, width) {
+    const keyOffset = offset + index * width;
+    const keyIndirectOffset = keyOffset - readInt(dataView, keyOffset, flexbuffers.BitWidth.fromByteWidth(width));
+    for (let i = 0; i < input.length; i++) {
+      const dif = input[i] - dataView.getUint8(keyIndirectOffset + i);
+      if (dif !== 0) {
+        return dif;
+      }
+    }
+    return dataView.getUint8(keyIndirectOffset + input.length) === 0 ? 0 : -1;
+  }
+
+  function valueForIndexWithKey(index, key, dataView, offset, parentWidth, byteWidth, length, path) {
+    const _indirect = indirect(dataView, offset, parentWidth);
+    const elementOffset = _indirect + index * byteWidth;
+    const packedType = dataView.getUint8(_indirect + length * byteWidth + index);
+    return Reference(dataView, elementOffset, flexbuffers.BitWidth.fromByteWidth(byteWidth), packedType, `${path}/${key}`)
+  }
+
+  function keyForIndex(index, dataView, offset, parentWidth, byteWidth) {
+    const keysVectorOffset = indirect(dataView, offset, parentWidth) - byteWidth * 3;
+    const bitWidth = flexbuffers.BitWidth.fromByteWidth(byteWidth)
+    const indirectOffset = keysVectorOffset - readInt(dataView, keysVectorOffset, bitWidth);
+    const _byteWidth = readInt(dataView, keysVectorOffset + byteWidth, bitWidth);
+    const keyOffset = indirectOffset + index * _byteWidth;
+    const keyIndirectOffset = keyOffset - readInt(dataView, keyOffset, flexbuffers.BitWidth.fromByteWidth(_byteWidth));
+    var length = 0;
+    while (dataView.getUint8(keyIndirectOffset + length) !== 0) {
+      length++;
+    }
+    return fromUTF8Array(new Uint8Array(dataView.buffer, keyIndirectOffset, length));
+  }
+
+  function Reference(dataView, offset, parentWidth, packedType, path) {
+    const byteWidth = 1 << (packedType & 3);
+    const valueType = packedType >> 2;
+    let length = -1;
+    return {
```

Comment:
This would be better done as a proper `class` to avoid constructing new objects and closures on each construction.